File: mobile/customer-app/app/src/main/java/com/sukashawarma/customer/ui/menu/CatalogViewModel.kt
package com.sukashawarma.customer.ui.menu

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sukashawarma.customer.data.CartStore
import com.sukashawarma.customer.data.OutletStore
import com.sukashawarma.customer.data.Repository
import com.sukashawarma.customer.data.api.BannerDto
import com.sukashawarma.customer.data.api.GatewayError
import com.sukashawarma.customer.data.api.GatewayResult
import com.sukashawarma.customer.data.api.MenuItemDto
import com.sukashawarma.customer.data.api.OutletDto
import com.sukashawarma.customer.ui.components.BottomNavTab
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

enum class CatalogScrollEvent { KeAtas, KeMenu }

data class CatalogState(
    val memuat: Boolean = true,
    val galat: GatewayError? = null,
    val outlet: OutletDto? = null,
    val kueri: String = "",
    val semuaItem: List<MenuItemDto> = emptyList(),
    val kategori: List<KategoriMenu> = emptyList(),
    /** Tidak ada outlet tersimpan dan tidak bisa ditentukan sendiri. */
    val perluPilihOutlet: Boolean = false,
    /** Tidak ada satu pun outlet yang ikut serta di aplikasi. */
    val tidakAdaOutlet: Boolean = false,
    /** Keranjang terhapus karena pelanggan berpindah outlet. */
    val keranjangDikosongkan: Boolean = false,
    val bannerCarousel: List<BannerDto> = emptyList(),
    val bannerPopup: BannerDto? = null
)

class CatalogViewModel(
    private val repository: Repository,
    private val outletStore: OutletStore,
    private val cart: CartStore
) : ViewModel() {

    private val _state = MutableStateFlow(CatalogState())
    val state: StateFlow<CatalogState> = _state.asStateFlow()

    private val _eventScroll = MutableSharedFlow<CatalogScrollEvent>(extraBufferCapacity = 1)
    val eventScroll: SharedFlow<CatalogScrollEvent> = _eventScroll.asSharedFlow()

    private val _activeTab = MutableStateFlow(BottomNavTab.BERANDA)
    val activeTab: StateFlow<BottomNavTab> = _activeTab.asStateFlow()

    fun setTab(tab: BottomNavTab) {
        _activeTab.value = tab
        when (tab) {
            BottomNavTab.BERANDA -> _eventScroll.tryEmit(CatalogScrollEvent.KeAtas)
            BottomNavTab.MENU -> _eventScroll.tryEmit(CatalogScrollEvent.KeMenu)
            else -> {}
        }
    }

    init {
        muat()
    }

    fun muat() {
        _state.value = _state.value.copy(memuat = true, galat = null)
        viewModelScope.launch {
            when (val hasil = repository.outlets()) {
                is GatewayResult.Gagal -> {
                    _state.value = _state.value.copy(memuat = false, galat = hasil.error)
                }
                is GatewayResult.Sukses -> {
                    // Banner tidak boleh menjatuhkan Beranda ATAU menahannya:
                    // gagal memuatnya berarti tidak ada banner, bukan layar
                    // galat, dan lambat memuatnya tidak boleh membuat seluruh
                    // layar menunggu -- dijalankan di coroutine terpisah
                    // (bukan di-await di jalur ini) supaya katalog, isi utama
                    // halaman ini, tidak duduk di spinner menunggu banner.
                    viewModelScope.launch {
                        val banner = repository.banners()
                        if (banner is GatewayResult.Sukses) {
                            _state.value = _state.value.copy(
                                bannerCarousel = banner.data.carousel,
                                bannerPopup = banner.data.popup
                            )
                        }
                    }

                    val outlets = hasil.data
                    if (outlets.isEmpty()) {
                        _state.value = _state.value.copy(
                            memuat = false,
                            tidakAdaOutlet = true
                        )
                        return@launch
                    }

                    // Outlet tersimpan yang sudah dicabut dari aplikasi tidak
                    // boleh tetap dipakai: pelanggan akan melihat menu outlet
                    // yang tak lagi melayani pesanan aplikasi.
                    val tersimpan = outletStore.idTerpilih()
                    val terpilih = outlets.firstOrNull { it.id == tersimpan }
                        ?: outlets.singleOrNull()

                    if (terpilih == null) {
                        _state.value = _state.value.copy(
                            memuat = false,
                            perluPilihOutlet = true
                        )
                        return@launch
                    }

                    outletStore.simpan(terpilih.id, terpilih.name)
                    pasangOutletKeKeranjang(terpilih.id)
                    muatKatalog(terpilih)
                }
            }
        }
    }

    fun pilihOutlet(outlet: OutletDto) {
        outletStore.simpan(outlet.id, outlet.name)
        pasangOutletKeKeranjang(outlet.id)
        _state.value = _state.value.copy(
            memuat = true,
            galat = null,
            perluPilihOutlet = false,
            outlet = outlet,
            semuaItem = emptyList(),
            kategori = emptyList()
        )
        viewModelScope.launch { muatKatalog(outlet) }
    }

    fun akuiKeranjangDikosongkan() {
        _state.value = _state.value.copy(keranjangDikosongkan = false)
    }

    /**
     * `menu_item_id` bersifat per-outlet, jadi keranjang milik outlet lain
     * pasti ditolak gateway saat checkout. Mengosongkannya di sini -- saat
     * pelanggan masih di katalog dan bisa memesan ulang -- jauh lebih baik
     * daripada membiarkannya gagal di titik pembayaran.
     */
    private fun pasangOutletKeKeranjang(outletId: String) {
        if (cart.pakaiOutlet(outletId)) {
            _state.value = _state.value.copy(keranjangDikosongkan = true)
        }
    }

    fun ubahKueri(kueri: String) {
        val s = _state.value
        _state.value = s.copy(
            kueri = kueri,
            kategori = kelompokkanPerKategori(saringPencarian(s.semuaItem, kueri))
        )
    }

    private suspend fun muatKatalog(outlet: OutletDto) {
        when (val hasil = repository.katalog(outlet.id)) {
            is GatewayResult.Gagal -> {
                // Outlet tetap dipasang walau katalog gagal: kepala layar dan
                // tombol "Ganti" harus tetap bisa dipakai, kalau tidak
                // pelanggan terjebak di layar galat tanpa jalan keluar.
                _state.value = _state.value.copy(
                    memuat = false,
                    outlet = outlet,
                    galat = hasil.error
                )
            }
            is GatewayResult.Sukses -> {
                val kueri = _state.value.kueri
                _state.value = _state.value.copy(
                    memuat = false,
                    galat = null,
                    outlet = outlet,
                    semuaItem = hasil.data,
                    kategori = kelompokkanPerKategori(saringPencarian(hasil.data, kueri))
                )
            }
        }
    }
}

File: mobile/customer-ios-app/SukaShawarma/UI/Menu/CatalogViewModel.swift
import Foundation
import Observation

/// Padanan `ui/menu/CatalogViewModel.kt`. Dimiliki shell navigasi (bukan tiap
/// layar), dipakai bersama tab Beranda & Menu: katalog yang sudah dimuat tidak
/// hilang saat pelanggan berpindah tab.
@MainActor
@Observable
final class CatalogViewModel {
    private(set) var memuat = true
    private(set) var galat: GatewayError?
    private(set) var outlet: OutletDto?
    var kueri = "" { didSet { if kueri != oldValue { susunKategori() } } }
    private(set) var semuaItem: [MenuItemDto] = []
    private(set) var kategori: [KategoriMenu] = []
    /// Ada >1 outlet, tak ada yang tersimpan — pelanggan harus memilih.
    private(set) var perluPilihOutlet = false
    /// Tidak ada satu pun outlet yang ikut serta di aplikasi.
    private(set) var tidakAdaOutlet = false
    /// Keranjang terhapus karena pelanggan berpindah outlet.
    private(set) var keranjangDikosongkan = false
    private(set) var bannerCarousel: [BannerDto] = []
    private(set) var bannerPopup: BannerDto?

    @ObservationIgnored private let repository: Repository
    @ObservationIgnored private let outletStore: OutletStore
    @ObservationIgnored private let cart: CartStore
    @ObservationIgnored private var tugasBanner: Task<Void, Never>?

    init(repository: Repository, outletStore: OutletStore, cart: CartStore) {
        self.repository = repository
        self.outletStore = outletStore
        self.cart = cart
    }

    func muat() async {
        memuat = true
        galat = nil
        perluPilihOutlet = false
        tidakAdaOutlet = false

        switch await repository.outlets() {
        case .gagal(let g):
            memuat = false
            galat = g
        case .sukses(let outlets):
            // Banner tidak boleh menjatuhkan ATAU menahan Beranda: dimuat
            // terpisah, gagal = tanpa banner, lambat = katalog tak menunggu.
            tugasBanner?.cancel()
            tugasBanner = Task { [repository] in
                if case .sukses(let b) = await repository.banners() {
                    bannerCarousel = b.carousel
                    bannerPopup = b.popup
                }
            }

            guard !outlets.isEmpty else {
                memuat = false
                tidakAdaOutlet = true
                return
            }

            // Outlet tersimpan yang sudah dicabut dari aplikasi tidak boleh tetap dipakai.
            let tersimpan = outletStore.idTerpilih()
            guard let terpilih = outlets.first(where: { $0.id == tersimpan })
                    ?? (outlets.count == 1 ? outlets[0] : nil)
            else {
                memuat = false
                perluPilihOutlet = true
                return
            }

            outletStore.simpan(id: terpilih.id, nama: terpilih.name)
            pasangOutletKeKeranjang(terpilih.id)
            await muatKatalog(terpilih)
        }
    }

    func pilihOutlet(_ o: OutletDto) async {
        outletStore.simpan(id: o.id, nama: o.name)
        pasangOutletKeKeranjang(o.id)
        memuat = true
        galat = nil
        perluPilihOutlet = false
        outlet = o
        semuaItem = []
        kategori = []
        await muatKatalog(o)
    }

    func akuiKeranjangDikosongkan() { keranjangDikosongkan = false }

    /// Keranjang outlet lain pasti ditolak gateway saat checkout — lebih baik
    /// dikosongkan di sini, selagi pelanggan masih bisa memesan ulang.
    private func pasangOutletKeKeranjang(_ outletId: String) {
        if cart.pakaiOutlet(outletId) { keranjangDikosongkan = true }
    }

    private func susunKategori() {
        kategori = kelompokkanPerKategori(saringPencarian(semuaItem, kueri: kueri))
    }

    private func muatKatalog(_ o: OutletDto) async {
        switch await repository.katalog(outletId: o.id) {
        case .gagal(let g):
            // Outlet tetap dipasang walau katalog gagal: tombol "Ganti" harus
            // tetap bisa dipakai, kalau tidak pelanggan terjebak.
            memuat = false
            outlet = o
            galat = g
        case .sukses(let items):
            memuat = false
            galat = nil
            outlet = o
            semuaItem = items
            susunKategori()
        }
    }
}
